pynear/include/BKTree.hpp

```diff
@@ -0,0 +1,156 @@
+    }
+
+    void clear() {
+        for (auto const &[d, bknode] : leaves) {
+            delete bknode;
+        }
+        leaves.clear();
+    }
+
+    virtual ~BKNode() { clear(); }
+};
+
+template <typename key_t, typename distance_t, typename metric> class BKTree {
+    BKNode<key_t, distance_t> *root;
+
+    public:
+    BKTree() : root(nullptr) {}
+
+    bool add(key_t key) {
+        if (root == nullptr) {
+            root = new BKNode<key_t, distance_t>(key);
+        } else {
+            BKNode<key_t, distance_t> *node = root;
+            distance_t dist;
+
+            while (true) {
+                dist = metric::distance(node->key, key);
+                auto next_it = node->leaves.find(dist);
+                if (next_it == node->leaves.end() || dist == 0) {
+                    break;
+                }
+                node = next_it->second;
+            }
+
+            if (dist > 0) {
+                node->add_leaf(dist, key);
+            } else {
+                return false; // didn't insert key
+            }
+        }
+        return true; // inserted new key
+    }
+
+    void update(std::vector<key_t> keys) {
+        for (auto const &key : keys) {
+            add(key);
+        }
+    }
+
+    std::tuple<std::vector<distance_t>, std::vector<key_t>> find(key_t key, distance_t threshold) {
+        static_assert(std::is_signed<distance_t>::value, "Arithmetic required signed distances");
+
+        BKNode<key_t, distance_t> *node = root;
+        std::vector<distance_t> distances;
+        std::vector<key_t> keys;
+
+        if (node == nullptr) {
+            return std::make_tuple(distances, keys);
+        }
+
+        std::deque<BKNode<key_t, distance_t> *> candidates = {node};
+        distance_t distance_cutoff, dist, lower, upper;
+        BKNode<key_t, distance_t> *candidate;
+        std::optional<distance_t> dist_opt;
+
+        while (!candidates.empty()) {
+            candidate = candidates.front();
+            candidates.pop_front();
+            distance_cutoff = candidate->max_distance.value_or(0) + threshold;
+            dist_opt = metric::threshold_distance(key, candidate->key, distance_cutoff);
+
+            if (!dist_opt.has_value()) {
+                continue;
+            }
+
+            dist = dist_opt.value();
+
+            if (dist <= threshold) {
+                distances.push_back(dist);
+                keys.push_back(candidate->key);
+            }
+
+            lower = dist - threshold;
+            upper = dist + threshold;
+            for (auto const &[d, bknode] : candidate->leaves) {
+                if (lower <= d && d <= upper) {
+                    candidates.push_back(bknode);
+                }
+            }
+        }
+        return std::make_tuple(distances, keys);
+    }
+
+    std::tuple<std::vector<std::vector<distance_t>>, std::vector<std::vector<key_t>>> find_batch(const std::vector<key_t> &keys,
```

Comment:
Later on we can add OPEN MP loop here like in vp tree. This greatly improves performance. But can be in a separate PR.